### ISSUE TITLE
Issue with sam build --build-in-source

### ISSUE THREAD

**Issue by aleguern-azivko** [aws/aws-sam-cli#7516]
When running sam build --build-in-source, the build process fails for the lambda function with an esbuild error. However, sam build without the --build-in-source flag works successfully.

**Steps to reproduce:**
Set up a SAM project with multiple lambda functions using esbuild.
Run sam build --build-in-source

**Expected behavior:**
The build process should complete successfully for all lambda functions.

**Actual behavior:**
```
The build fails with the following error:
CopyNodejsNpmEsbuildBuilder:EsbuildBundle failed
EsbuildExecutionError: Esbuild Failed: ✘ [ERROR] Could not resolve "@sendgrid/client"

node_modules/@sendgrid/mail/src/classes/mail-service.js:6:25:
  6 │ const {Client} = require('@sendgrid/client');
    ╵                          ~~~~~~~~~~~~~~~~~~

You can mark the path "@sendgrid/client" as external to exclude it from the bundle, which will remove this error. You can also surround this "require" call with a try/catch block to handle this failure at run-time instead of bundle-time.
```

**Environment:**
npm version: 10.5.0
SAM CLI version: 1.124.0 
Operating System: macOS

**Relevant configuration:**
SAM template snippet for the Contact function:

```
  Type: AWS::Serverless::Function
  Properties:
    Environment:
      Variables:
        SENDGRID_API_KEY: !Ref SendGridApiKey
    Events:
      Contact:
        Type: Api
        Properties:
          RestApiId: !Ref Api
          Path: /contact
          Method: post
          Auth:
            Authorizer: NONE
  Metadata:
    BuildMethod: esbuild
    BuildProperties:
      Minify: true
      Target: 'es2022'
      Sourcemap: false
      EntryPoints:
        - src/public/contact/index.ts
```

**Relevant dependencies from package.json:**
```
"dependencies": {
  "@sendgrid/mail": "^7.7.0",
  "esbuild": "^0.17.17"
},
"devDependencies": {
  "@types/aws-lambda": "^8.10.145",
  "typescript": "^4.5.5"
}
```

**Comment by hawflau** [comment]
@aleguern-azivko thanks for raising issue and for finding the root cause. If build-in-source doesn't work well with parallel, I believe we should check and not allow them to be active at the same time. Marking it as a bug

**Comment by aleguern-azivko** [comment]
To add a little bit of context, the bug occurred using typescript and npm, didn’t try other languages.

**Comment by dcabib** [comment]
Hey, @hawflau @aleguern-azivko managed to reproduce this locally. Here's the breakdown:

**What's going on**

When you use `--build-in-source` + `--parallel` and your lambdas share the same `CodeUri`, multiple npm processes end up running on the same node_modules directory at once:

```
Running NodejsNpmEsbuildBuilder:NpmUpdate
Running NodejsNpmEsbuildBuilder:NpmUpdate  
Running NodejsNpmEsbuildBuilder:NpmUpdate
```

Classic race condition - npm processes fighting over the same dir = corrupted esbuild binary. The error is intermittent since it depends on timing.

**Proposed fix**

Blocking the flag combination would be the easy way out but it limits functionality. A better approach would be adding per-codeuri locking to `ParallelBuildStrategy`.

The idea:
- When `build_in_source=True`, acquire a lock based on the codeuri before building
- Same codeuri = wait your turn (sequential)
- Different codeuris = still run in parallel (no perf impact for unrelated functions)

The changes would touch:
- `samcli/lib/build/build_strategy.py` - locking logic in `ParallelBuildStrategy`
- `samcli/lib/build/app_builder.py` - pass `build_in_source` to the strategy

I prototyped this locally and tested with a template with 3 Lambda functions sharing the same CodeUri (with esbuild, @sendgrid/mail dependency). Builds consistently succeed now without the binary corruption.

Happy to open a PR if this approach makes sense.

**Comment by sxflynn** [comment]
@dcabib  I am very interested in your bug fix and look forward to you submitting a PR one day and getting it merged into the cli.
